Src/VimCore/CaretRegisterMap.fs

```diff
@@ -0,0 +1,123 @@
+﻿namespace Vim
+
+open System.Collections.Generic
+
+/// Mutable caret index object
+type internal CaretIndex() =
+    let mutable _caretIndex = 0
+    member x.Value
+        with get() = _caretIndex
+        and set value = _caretIndex <- value
+
+/// Caret-aware IRegisterValueBacking implementation for the unnamed register.
+/// If there are multiple carets, each caret gets its own distinct value
+type internal CaretUnnamedRegisterValueBacking
+    (
+        _caretIndex: CaretIndex,
+        _unnamedRegister: Register
+    ) =
+
+    let _valueMap = Dictionary<int, RegisterValue>()
+
+    member x.RegisterValue = 
+        if _caretIndex.Value = 0 then
+            _unnamedRegister.RegisterValue
+        else
+            match _valueMap.TryGetValue(_caretIndex.Value) with
+            | true, value -> value
+            | _ -> _unnamedRegister.RegisterValue
+
+    member x.SetRegisterValue value =
+        if _caretIndex.Value = 0 then
+            _unnamedRegister.RegisterValue <- value
+        else
+            _valueMap.[_caretIndex.Value] <- value
+
+    interface IRegisterValueBacking with
+        member x.RegisterValue 
+            with get () = x.RegisterValue
+            and set value = x.SetRegisterValue value
+
+/// Caret-aware IRegisterValueBacking implementation for the clipboard
+/// register. If there are multiple carets, the primary caret uses the
+/// clipboard and each secondary caret uses that caret's unnamed register
+type internal CaretClipboardRegisterValueBacking
+    (
+        _caretIndex: CaretIndex,
+        _clipboardRegister: Register,
+        _caretUnnamedRegister: Register
+    ) =
+    member x.RegisterValue = 
+        if _caretIndex.Value = 0 then
+            _clipboardRegister.RegisterValue
+        else
+            _caretUnnamedRegister.RegisterValue
+
+    member x.SetRegisterValue (value: RegisterValue) =
+        if _caretIndex.Value = 0 then
+            _clipboardRegister.RegisterValue <- value
+        else
+            _caretUnnamedRegister.RegisterValue <- value
+
+    interface IRegisterValueBacking with
+        member x.RegisterValue 
+            with get () = x.RegisterValue
+            and set value = x.SetRegisterValue value
+
+type CaretRegisterMap

```

Comment:
This should be `type internal` to match the definition in the FSI file. I'm a bit surprised that F# doesn't warn about this case. Not sure which wins on emit but I'm guessing that it's the version in the .fs file.